internal/statesync/stateprovider.go

```diff
@@ -203,35 +318,66 @@ func (s *lightClientStateProvider) State(ctx context.Context, height uint64) (sm
 	state.NextValidators = nextLightBlock.ValidatorSet
 	state.LastHeightValidatorsChanged = nextLightBlock.Height
 
-	// We'll also need to fetch consensus params via RPC, using light client verification.
-	primaryURL, ok := s.providers[s.lc.Primary()]
-	if !ok || primaryURL == "" {
-		return sm.State{}, fmt.Errorf("could not find address for primary light client provider")
-	}
-	primaryRPC, err := rpcClient(primaryURL)
+	// We'll also need to fetch consensus params via P2P.
+	state.ConsensusParams, err = s.consensusParams(ctx, currentLightBlock.Height)
 	if err != nil {
-		return sm.State{}, fmt.Errorf("unable to create RPC client: %w", err)
+		return sm.State{}, err
 	}
-	rpcclient := lightrpc.NewClient(primaryRPC, s.lc)
-	result, err := rpcclient.ConsensusParams(ctx, &currentLightBlock.Height)
-	if err != nil {
-		return sm.State{}, fmt.Errorf("unable to fetch consensus parameters for height %v: %w",
-			nextLightBlock.Height, err)
+	// validate the consensus params
+	if !bytes.Equal(nextLightBlock.ConsensusHash, state.ConsensusParams.HashConsensusParams()) {
+		return sm.State{}, fmt.Errorf("consensus params hash mismatch at height %d. Expected %v, got %v",
+			currentLightBlock.Height, nextLightBlock.ConsensusHash, state.ConsensusParams.HashConsensusParams())
 	}
-	state.ConsensusParams = result.ConsensusParams
+	// set the last height changed to the current height
 	state.LastHeightConsensusParamsChanged = currentLightBlock.Height
 
 	return state, nil
 }
 
-// rpcClient sets up a new RPC client
-func rpcClient(server string) (*rpchttp.HTTP, error) {
-	if !strings.Contains(server, "://") {
-		server = "http://" + server
+// addProvider dynamically adds a peer as a new witness. A limit of 6 providers is kept as a
+// heuristic. Too many overburdens the network and too little compromises the second layer of security.
+func (s *stateProviderP2P) addProvider(p lightprovider.Provider) {
+	if len(s.lc.Witnesses()) < 6 {
+		s.lc.AddProvider(p)
 	}
-	c, err := rpchttp.New(server)
-	if err != nil {
-		return nil, err
+}
+
+// consensusParams sends out a request for consensus params blocking until one is returned.
+// If it fails to get a valid set of consensus params from any of the providers it returns an error.
+func (s *stateProviderP2P) consensusParams(ctx context.Context, height int64) (types.ConsensusParams, error) {
+	providers := s.lc.Witnesses()
+	for _, provider := range providers {
+		p, ok := provider.(*BlockProvider)
+		if !ok {
+			panic("expected p2p state provider to use p2p block providers")
+		}
+
+		// extract the nodeID of the provider
+		peer, err := types.NewNodeID(p.String())
+		if err != nil {
+			return types.ConsensusParams{}, fmt.Errorf("invalid provider node id: %w, provider: %s", err, p.String())
```

Comment:
Cosmetic suggestion:

```go
fmt.Errorf("invalid provider %s node id: %w", p.String(), err)
```

just so that underlying errors unwind to the right. Not a big deal.